src/AlignmentFixerHelper.php

```diff
@@ -0,0 +1,341 @@
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer;
+
+use PhpCsFixer\Tokenizer\CT;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author SpacePossum
+ *
+ * @internal
+ */
+final class AlignmentFixerHelper
+{
+    /**
+     * @internal
+     */
+    const ALIGN = 'align';
+
+    /**
+     * @internal
+     */
+    const ALIGN_SINGLE_SPACE = 'align_single_space';
+
+    /**
+     * @internal
+     */
+    const ALIGN_SINGLE_SPACE_MINIMAL = 'align_single_space_minimal';
+
+    /**
+     * @const Placeholder used as anchor for right alignment.
+     */
+    const ALIGNABLE_PLACEHOLDER = "\x2 ALIGNABLE%d \x3";
+
+    /**
+     * Keep track of the deepest level ever achieved while
+     * parsing the code. Used later to replace alignment
+     * placeholders with spaces.
+     *
+     * @var int
+     */
+    private $deepestLevel = 0;
+
+    /**
+     * Level counter of the current nest level.
+     * So one level alignments are not mixed with
+     * other level ones.
+     *
+     * @var int
+     */
+    private $currentLevel = 0;
+
+    /**
+     * @param Tokens                $tokens
+     * @param array<string, string> $toAlign
+     */
+    public function fixAlignment(Tokens $tokens, array $toAlign)
+    {
+        foreach ($toAlign as $tokenContent => $alignStrategy) {
+            if (self::ALIGN !== $alignStrategy && self::ALIGN_SINGLE_SPACE !== $alignStrategy && self::ALIGN_SINGLE_SPACE_MINIMAL !== $alignStrategy) {
+                throw new \UnexpectedValueException('Unknown strategy provided.');
+            }
+
+            // This fixer works partially on Tokens and partially on string representation of code.
+            // During the process of fixing internal state of single Token may be affected by injecting ALIGNABLE_PLACEHOLDER to its content.
+            // The placeholder will be resolved by `replacePlaceholder` method by removing placeholder or changing it into spaces.
+            // That way of fixing the code causes disturbances in marking Token as changed - if code is perfectly valid then placeholder
+            // still be injected and removed, which will cause the `changed` flag to be set.
+            // To handle that unwanted behavior we work on clone of Tokens collection and then override original collection with fixed collection.
+            $tokensClone = clone $tokens;
+
+            if ('=>' === $tokenContent) {
+                $this->injectAlignmentPlaceholdersForArrow($tokensClone, 0, count($tokens));
+            } else {
+                $this->injectAlignmentPlaceholders($tokensClone, 0, count($tokens), $tokenContent);
+            }
+
+            // for all tokens that should be aligned but do not have anything to align with, fix spacing if needed
+            if (self::ALIGN_SINGLE_SPACE === $alignStrategy || self::ALIGN_SINGLE_SPACE_MINIMAL === $alignStrategy) {
+                if ('=>' === $tokenContent) {
```

Comment:
IMO the two `foreach` can be merged to simplify code and improve readability. The second one already performs `strtolower()` calls for operators that don't need it anyway.